### PR TITLE
Move client specific python modules to avalon_client

### PR DESCRIPTION
**Pull request by Ram-srini**
1. Create a new python package for client sdk
2. Update the avalon client sdk interface as per the EEA spec changes
3. Client sdk package depends examples/common/python module
3. Added generic test into ci run script

Signed-off-by: Ramakrishna Srinivasamurthy <ramakrishna.srinivasamurthy@intel.com>

### CODE REVIEW

File: examples/common/python/database/lmdb_helper_proxy.py [hunk 1]
```diff
@@ -167,3 +175,60 @@ def __escape(self, string):
 # ------------------------------------------------------------------------------
     def __unescape(self, string):
         return string.encode("utf-8").decode("unicode_escape")
+
+
+class TextServiceClient(object):
```
Comment (manojgop) [inline examples/common/python/database/lmdb_helper_proxy.py]: This change (moving TextServiceClient to this file) should be separate PR

File: scripts/eth_cli.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import toml
+import errno
+import logging
+import os
+from os.path import exists, realpath
+
+from connectors.ethereum.ethereum_wrapper \
```
Comment (manojgop) [inline scripts/eth_cli.py]: connectors is renamed in this PR ?

File: scripts/eth_cli.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        contract_id, contract_interface = compiled_sol.popitem()
+        address = self.__eth_client.deploy_contract(contract_interface)
+        logging.info("Deployed %s to: %s\n", contract_id, address)
+        return {"status": "success"}
+
+
+def main():
+    tcf_home = os.environ.get("TCF_HOME", "../../")
+    eth = eth_cli(tcf_home +
+                  "/examples/common/python/connectors/" + "tcf_connector.toml")
```
Comment (manojgop) [inline scripts/eth_cli.py]: Need to change the path. Is this file tested ?

File: avalon_client/avalon_client_sdk/direct/jrpc/work_order_jrpc_client.py [hunk 4]
```diff
@@ -88,3 +82,40 @@ def work_order_get_result(self, work_order_id, id=None):
                     # responseTimeoutMsecs in the request.
                     time.sleep(2)
                 return response
+
+    def encryption_key_retrieve(self, worker_id, last_used_key_nonce, tag,
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply avalon_client/avalon_client_sdk/direct/jrpc/work_order_jrpc_client.py]: encryption_key_retrieve is existing feature and updated the name as per the latest spec in proxy model and encryption_key_start  is newly added in spec so added in this PR.

File: avalon_client/avalon_client_sdk/direct/jrpc/work_order_receipt_jrpc_client.py [hunk 5]
```diff
@@ -16,17 +16,17 @@
...
-from service_client.generic import GenericServiceClient
-from connectors.interfaces.work_order_receipt_interface \
-    import WorkOrderReceiptInterface
-from utility.tcf_types import ReceiptCreateStatus
-from connectors.utils import create_jrpc_response, validate_details
+from avalon_client_sdk.service_client.generic import GenericServiceClient
+from avalon_client_sdk.interfaces.work_order_receipt_client \
+    import WorkOrderReceiptClient
+from avalon_client_sdk.utility.tcf_types import ReceiptCreateStatus
+from avalon_client_sdk.utility.utils import create_jrpc_response
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply avalon_client/avalon_client_sdk/direct/jrpc/work_order_receipt_jrpc_client.py]: I guess JSONRPCDispatchException has to be called from handler function, anyway i have to remove it from client because client shouldn't through any exception.

File: avalon_client/avalon_client_sdk/direct/jrpc/worker_registry_jrpc_client.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
+# Copyright 2019 Intel Corporation
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply avalon_client/avalon_client_sdk/direct/jrpc/worker_registry_jrpc_client.py]: existing file worker_registry_jrpc_impl.py is renamed to worker_registry_jrpc_client.py

File: avalon_client/avalon_client_sdk/direct/direct_ethereum_client.py [hunk 7]
```diff
@@ -17,19 +17,19 @@
...
 
-from connectors.interfaces.connector_interface import \
-    ConnectorInterface
-from connectors.ethereum.ethereum_worker_registry_list_impl import \
-    EthereumWorkerRegistryListImpl
-from connectors.direct.worker_registry_jrpc_impl import WorkerRegistryJRPCImpl
-from connectors.direct.work_order_jrpc_impl import WorkOrderJRPCImpl
-from connectors.direct.work_order_receipt_jrpc_impl \
-     import WorkOrderReceiptJRPCImpl
+from avalon_client_sdk.interfaces.client import \
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply avalon_client/avalon_client_sdk/direct/direct_ethereum_client.py]: Avalon client APIs are called by the requester/client like submitting work order, getting work order result, worker lookup, etc.. these APIs are not being called by Connector/worker service. Connector/work service will call register worker, setting worker status, updating worker, creating work order response etc.

File: avalon_client/avalon_client_sdk/interfaces/client.py [hunk 8]
```diff
@@ -15,7 +15,7 @@
 from abc import ABC, abstractmethod
 
 
-class ConnectorInterface(ABC):
+class Client(ABC):
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply avalon_client/avalon_client_sdk/interfaces/client.py]: Yes Avalon client is different from Connector. Name client represent it is client/requester.

File: avalon_client/avalon_client_sdk/direct/direct_ethereum_client.py [hunk 9]
```diff
@@ -17,19 +17,19 @@
...
+from avalon_client_sdk.direct.jrpc.worker_registry_jrpc_client import WorkerRegistryJRPCClient
+from avalon_client_sdk.direct.jrpc.work_order_jrpc_client import WorkOrderJRPCClient
+from avalon_client_sdk.direct.jrpc.work_order_receipt_jrpc_client \
+     import WorkOrderReceiptJRPCClient
 
 logger = logging.getLogger(__name__)
 
 
-class DirectJsonRpcApiConnector(ConnectorInterface):
+class DirectEthereumClient(Client):
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply avalon_client/avalon_client_sdk/direct/direct_ethereum_client.py]: Client meaning clients are using these classes.

File: scripts/eth_cli.py (hunk 2 above)
Comment (Ram-srini) [reply scripts/eth_cli.py]: Yes it has to be updated.

File: scripts/eth_cli.py (hunk 3 above)
Comment (Ram-srini) [reply scripts/eth_cli.py]: yes i need to change it.

File: examples/common/python/database/lmdb_helper_proxy.py (hunk 1 above)
Comment (Ram-srini) [reply examples/common/python/database/lmdb_helper_proxy.py]: Anyway it has to be separated out and this PR is separate our client specific modules. Client won't use it and specific lmdb_helper_proxy.

File: avalon_client/avalon_client_sdk/direct/direct_ethereum_client.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (manojgop) [reply avalon_client/avalon_client_sdk/direct/direct_ethereum_client.py]: Can we rename the connector class with Connector suffix if we plan to keep it in avalon_client_sdk which will be inline with the design/class diagrams.

File: avalon_client/avalon_client_sdk/direct/direct_ethereum_client.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (manojgop) [reply avalon_client/avalon_client_sdk/direct/direct_ethereum_client.py]: Let's remove Client class and use AvalonDirectClient for all blockchain types. If blockchain type is none , read uri from config file. AvalonDirectClient class constructor will create instance of WorkerRegistryListClient, WorkerRegistryClient ,  WorkOrderClient and WorkOrderReceiptClient

File: avalon_client/avalon_client_sdk/direct/direct_ethereum_client.py [hunk 10]
```diff
@@ -71,21 +71,21 @@ def __init__(self, config_file=None, config=None):
     def create_worker_registry_list(self, config):
         if self.__blockchain_type == "Ethereum":
             if self.__worker_registry_list is None:
-                self.__worker_registry_list = EthereumWorkerRegistryListImpl(
+                self.__worker_registry_list = EthereumWorkerRegistryListClient(
```
Comment (manojgop) [inline avalon_client/avalon_client_sdk/direct/direct_ethereum_client.py]: EthereumWorkerRegistryListClient -> EthereumWorkerRegistryListClientImpl

File: avalon_client/avalon_client_sdk/interfaces/work_order_client.py [hunk 11]
```diff
@@ -15,7 +15,7 @@
 from abc import ABC, abstractmethod
 
 
-class WorkOrderInterface(ABC):
+class WorkOrderClient(ABC):
     """
     WorkOrderRegistryInterface is an abstract base class that contains
```
Comment (manojgop) [inline avalon_client/avalon_client_sdk/interfaces/work_order_client.py]: WorkOrderRegistryInterface  -> This is an

File: avalon_client/avalon_client_sdk/interfaces/work_order_receipt_client.py [hunk 12]
```diff
@@ -15,7 +15,7 @@
 from abc import ABC, abstractmethod
 
 
-class WorkOrderReceiptInterface(ABC):
+class WorkOrderReceiptClient(ABC):
     """
     WorkOrderReceiptInterface is an abstract base class that contains
```
Comment (manojgop) [inline avalon_client/avalon_client_sdk/interfaces/work_order_receipt_client.py]: WorkOrderReceiptInterface  is an -> "This is an"

File: avalon_client/avalon_client_sdk/interfaces/work_order_receipt_client.py [hunk 13]
```diff
@@ -179,6 +185,9 @@ def work_order_receipt_lookup(self, worker_service_id,
         parameter.
         ids is an array of the Work Order receipt ids that match the input
         parameters.
+        Returns
+            tupe containing matching count, lookup tag and list of work
```
Comment (manojgop) [inline avalon_client/avalon_client_sdk/interfaces/work_order_receipt_client.py]: tupe -> tuple

File: avalon_client/avalon_client_sdk/interfaces/worker_registry_client.py [hunk 14]
```diff
@@ -15,7 +15,7 @@
 from abc import ABC, abstractmethod
 
 
-class WorkerRegistryInterface(ABC):
+class WorkerRegistryClient(ABC):
     """
     WorkerRegistryInterface is an abstract base class containing abstract APIs
```
Comment (manojgop) [inline avalon_client/avalon_client_sdk/interfaces/worker_registry_client.py]: WorkerRegistryInterface is an  -> This is an

File: avalon_client/avalon_client_sdk/interfaces/worker_registry_list_client.py [hunk 15]
```diff
@@ -14,61 +14,11 @@
 
 from abc import ABC, abstractmethod
 
-
-class WorkerRegistryListInterface(ABC):
+class WorkerRegistryListClient(ABC):
```
Comment (manojgop) [inline avalon_client/avalon_client_sdk/interfaces/worker_registry_list_client.py]: Class description is missing

File: avalon_client/avalon_client_sdk/work_order/__init__.py [hunk 16]
```diff
@@ -0,0 +1,15 @@
+# Copyright 2019 Intel Corporation
```
Comment (manojgop) [inline avalon_client/avalon_client_sdk/work_order/__init__.py]: 2019 -> 2020

File: avalon_client/setup.py [hunk 17]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env python
+
+# Copyright 2019 Intel Corporation
```
Comment (manojgop) [inline avalon_client/setup.py]: 2019 -> 2020

Comment (manojgop) [reply avalon_client/setup.py]: Can we rename avalon_client directory to avalon_client_sdk. Do we need avalon_client directory ?

File: examples/apps/generic_client/generic_client.py [hunk 18]
```diff
@@ -166,7 +166,7 @@ def Main(args=None):
                 " *****************")
 
     global direct_jrpc
-    direct_jrpc = DirectJsonRpcApiConnector(config_file=None, config=config)
+    direct_jrpc = AvalonDirectClient(config_file=None, config=config)
```
Comment (manojgop) [inline examples/apps/generic_client/generic_client.py]: What's difference between config_file and config. Please add a comment in code..

File: examples/apps/generic_client/generic_client.py [hunk 19]
```diff
@@ -176,7 +176,7 @@ def Main(args=None):
     # Connect to registry list and retrieve registry
     global uri
     if not uri and mode == "listing":
-        registry_list_instance = direct_jrpc.create_worker_registry_list(
+        registry_list_instance = direct_jrpc.get_worker_registry_list_instance(
             config
```
Comment (manojgop) [inline examples/apps/generic_client/generic_client.py]: Do we need to pass config parameter again in this API since we are already passing this parameter in constructor ?

File: avalon_client/avalon_client_sdk/direct/direct_ethereum_client.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Ram-srini) [reply avalon_client/avalon_client_sdk/direct/direct_ethereum_client.py]: This is not connector, it is client API.

File: avalon_client/avalon_client_sdk/direct/direct_ethereum_client.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Ram-srini) [reply avalon_client/avalon_client_sdk/direct/direct_ethereum_client.py]: It is already removed.

File: avalon_client/avalon_client_sdk/direct/direct_ethereum_client.py (hunk 10 above)
Comment (Ram-srini) [reply avalon_client/avalon_client_sdk/direct/direct_ethereum_client.py]: Done

File: avalon_client/avalon_client_sdk/interfaces/worker_registry_client.py (hunk 14 above)
Comment (Ram-srini) [reply avalon_client/avalon_client_sdk/interfaces/worker_registry_client.py]: Done

File: avalon_client/avalon_client_sdk/interfaces/work_order_receipt_client.py (hunk 13 above)
Comment (Ram-srini) [reply avalon_client/avalon_client_sdk/interfaces/work_order_receipt_client.py]: Done

File: avalon_client/avalon_client_sdk/interfaces/worker_registry_list_client.py (hunk 15 above)
Comment (Ram-srini) [reply avalon_client/avalon_client_sdk/interfaces/worker_registry_list_client.py]: Added

File: avalon_client/setup.py (hunk 17 above)
Comment (Ram-srini) [reply avalon_client/setup.py]: Yes, to package all of the python modules into avalon_client_sdk package.

File: examples/apps/generic_client/generic_client.py (hunk 19 above)
Comment (Ram-srini) [reply examples/apps/generic_client/generic_client.py]: Not required

File: avalon_client/avalon_client_sdk/direct/jrpc/work_order_jrpc_client.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (manojgop) [reply avalon_client/avalon_client_sdk/direct/jrpc/work_order_jrpc_client.py]: I can see "encryptionKeyRetrieve" only in proxy model in TC spec. Direct model only have EncryptionKeyGet. So why is encryption_key_retrieve function present in this file which is for direct model ? We should not be supporting this function for direct model.

File: avalon_client/setup.py (hunk 17 above)
Comment (manojgop) [reply avalon_client/setup.py]: avalon_client can have a different interpretation like clients for avalon like echo_clinet, generic_client, heart_disease_eval etc. May be we can rename top level directory as avalon_client_sdk and inside that we can create an sdk directory for packaging python modules. or top level directory as sdk and inside that avalon_client_sdk and avalon_connector_sdk.

File: examples/apps/heart_disease_eval/client/heart_gui.py [hunk 20]
```diff
@@ -323,15 +323,16 @@ def evaluate(self, message):
 
         # Set text for JSON sidebar
         req_id = 51
-        self.request_json = wo_params.to_string(req_id)
+        self.request_json = wo_params.to_jrpc_string(req_id)
 
-        work_order_instance = direct_jrpc.create_work_order(
+        work_order_instance = direct_jrpc.get_work_order_instance(
             config
```
Comment (manojgop) [inline examples/apps/heart_disease_eval/client/heart_gui.py]: Do we need to pass config parameter in this function ? We can pass config in the constructor of direct_jrpc.

File: examples/apps/generic_client/generic_client.py (hunk 18 above)
Comment (Ram-srini) [reply examples/apps/generic_client/generic_client.py]: Comments are there in AvalonDirectClient

File: examples/apps/heart_disease_eval/client/heart_gui.py (hunk 20 above)
Comment (Ram-srini) [reply examples/apps/heart_disease_eval/client/heart_gui.py]: Not required

File: examples/apps/generic_client/generic_client.py (hunk 18 above)
Comment (manojgop) [inline examples/apps/generic_client/generic_client.py]: Opened an issue to implement Generic client using Avalon SDK APIs for worker registry, work order etc insated of using facade class -> https://github.com/hyperledger/avalon/issues/207
